janusgraph-core/src/main/java/org/janusgraph/graphdb/util/ProfiledIterator.java

```diff
@@ -0,0 +1,56 @@
+// Copyright 2021 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.util;
+
+import org.apache.tinkerpop.gremlin.structure.util.CloseableIterator;
+import org.janusgraph.core.JanusGraphElement;
+import org.janusgraph.graphdb.query.profile.QueryProfiler;
+
+import java.util.Iterator;
+import java.util.function.Supplier;
+
+/**
+ * @author Boxuan Li (liboxuan@connect.hku.hk)
+ */
+public class ProfiledIterator<E extends JanusGraphElement> extends CloseableAbstractIterator<E> {
+    private final QueryProfiler profiler;
+    private Iterator<E> iterator;
+    private boolean timerRunning;
+
+    public ProfiledIterator(QueryProfiler profiler, Supplier<Iterator<E>> iteratorSupplier) {
+        this.profiler = profiler;
+        profiler.startTimer();
+        timerRunning = true;
+        iterator = iteratorSupplier.get();
```

Comment:
(Nitpick) One more thing: I think you can move these three lines to `computeNext()`. In `computeNext()` we can then start the timer and construct the iterator if `timerRunning == false`.
By doing so, the timer is automatically started, completely independent of when the `ProfiledIterator` is created. One can then call the constructor at any time, but the measurement only starts once the iterator is actually used.

Reply:
In fact, I think it actually makes some sense to start the timer in the construction phase, since it means the control flow has entered the scope of ProfiledIterator.

Moving it to computeNext() method means an additional instance variable and an extra if statement for every computeNext() call.

After all, there is very little profile timing & overhead difference between  both ways.

Reply:
I agree that there is almost no difference in this case. I just thought about a case, where one would for example prepare multiple iterators and then use them one by one at some point in time later on. but maybe that use case is a little bit constructed and far from real world application.

Reply:
That actually sounds like a legitimate case. I have changed the logic.

Reply:
I had to revert this piece of logic, otherwise, JanusGraphIndexTest::testOrForceIndexPartialIndex would fail. The reason is when we put the "iteratorSupplier.get()" logic into the computeNext() method, the computation is lazy (delayed). In that test case, if the computation is delayed, there will be no exception for "g.V().or(__.has("name", "Totoro"),__.has("age", 2)).hasNext();" because the first condition uses index and returns a result (thus, the second condition is not yet evaluated). On the contrary, if the computation is not delayed, the second condition will throw an exception because it does not use an index, and force-index is enabled.

In fact, I think this actually demonstrates the benefit of lazy computation, but since this would change the behavior and can have a non-trivial impact, I would rather leave it as a follow-up issue (to discuss further whether we shall do so or not).

Reply:
So if I understand correctly, using lazy computation would actually cause an even better usage of indexes? Or does it cause a problem (in the sense of incorrect results), that the second condition is not evaluated in this case?
If it really is _that easy_ to enable index usage for more expressions, we should definitely elaborate further in the future.
If you stick with this solution for now, you can remove the `Supplier` and pass the `Iterator` directly.

Reply:
> using lazy computation would actually cause an even better usage of indexes? Or does it cause a problem (in the sense of incorrect results), that the second condition is not evaluated in this case?

It does not lead to "better usage of indexes" per se. What happens is, `g.V().or(has("name", "Totoro"),has("age", 2)).hasNext();` consists of two graph-centric queries, each wrapped by a "ProfiledIterator". The two "ProfiledIterator"s are then combined into a "MultiDistinctUnorderedIterator". The first condition returns non-empty results, thus "hasNext()" returns true without the need of evaluating the second condition.

> does it cause a problem (in the sense of incorrect results)

Yes and no. It does change the current behaviour, but I tend to believe this is not incorrect. Let's use the same example ("name" is indexed, "age" is not indexed, “hasForceIndexUsage" is true). The current behaviour is, both `g.V().or(has("name", "Totoro"),has("age", 2)).hasNext();`  and `g.V().or(has("age", 2),has("name", "Totoro")).hasNext();` lead to exception because no index is used. If we do this "lazy computation", then `g.V().or(has("age", 2),has("name", "Totoro")).hasNext();` leads to exception but `g.V().or(has("name", "Totoro"),has("age", 2)).hasNext();` not.

Reply:
> Nope. As we discussed earlier, to do that, I have to construct ProfiledIterator and start the timer first. 
Oh yes, my bad.

> If we do this "lazy computation", then g.V().or(has("age", 2),has("name", "Totoro")).hasNext(); leads to exception but g.V().or(has("name", "Totoro"),has("age", 2)).hasNext(); not.
Ok I understand. So while it sounds good to me in general to implement lazy computation, we would also have to ensure that the order of sub queries inside an `or` does not affect the outcome, i.e. both use the indexed part first.

I see that this needs a closer look at least and we should postpone it to another PR. Would you please create an issue for that?

Reply:
Created https://github.com/JanusGraph/janusgraph/issues/2407